### PR TITLE
[feat] added NgComponent decorator

### PR DESCRIPTION
**Pull request by jmlopez-rod**
Added component decorator so that we can register angularJS components
in a similar fashion as angular components.

### CODE REVIEW

File: example/src/ex-component.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  move(element: JQuery): void {
+    super.move(element);
+    this.$log.log('Component move called. Injection: ', this.exService);
+  }
+
+  $onInit(): void {
+    super.$onInit();
+    this.$element.css('color', 'red');
+  }
+}
```
Comment (RedStarThrower) [inline example/src/ex-component.ts:62]: Should there be a space before the closing bracket of a class?

File: src/ng-register.ts [hunk 2]
```diff
@@ -2,6 +2,15 @@
...
 import * as ng from 'angular';
+import { INgComponentOptions } from './component';
+
+function toCamelCase(str: string, pascal: boolean = false): string {
+  const words = str.split('-').map(word => word.charAt(0).toUpperCase() + word.slice(1));
+  if (!pascal) {
+    words[0] = words[0].toLowerCase();
+  }
+  return words.join('');
+}
```
Comment (RedStarThrower) [inline src/ng-register.ts:13]: This is pretty cool! No one can mess up registering their components now

File: example/src/ex-controller.ts [hunk 3]
```diff
@@ -5,19 +5,14 @@
 import { Directive, Inject, mix } from '../../src/index';
 import ExService from './ex-service';
 
-
 class Swimmer {
-
   swim(): void {
     console.log('swimming...');
   }
```
Comment (RedStarThrower) [inline example/src/ex-controller.ts:11]: You just keep swimming little buddy!!

File: src/component.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  templateUrl?: string | Injectable<(...args: any[]) => string>;
+  bindings?: {[boundProperty: string]: string};
+  transclude?: boolean | {[slot: string]: string};
+  require?: {[controller: string]: string};
+}
+
+/**
+ * Attaches `__component` to the clazz object. If it already has one, then it overrides the
+ * properties it needs to.
+ *
```
Comment (RedStarThrower) [inline src/component.ts]: I'm not sure I understand this comment. What overrides what here? Vague pronoun "it" makes it hard to know. Does the "attach" method override the properties existing on the clazz object?

File: src/component.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+function NgComponent(options: INgComponentOptions): Function {
+  return (target: any) => {
+    attach(target, options);
+  };
+}
+
+/**
+ * Called on each controller after all the controllers on an element have been constructed and had
+ * their bindings initialized (and before the pre & post linking functions for the directives on
+ * this element). This is a good place to put initialization code for your controller.
```
Comment (RedStarThrower) [inline src/component.ts:65]: Does an element sometimes have multiple controllers?

File: src/component.ts [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+/**
+ * Called after this controller's element and its children have been linked. Similar to the
+ * post-link function this hook can be used to set up DOM event handlers and do direct DOM
+ * manipulation. Note that child elements that contain templateUrl directives will not have been
+ * compiled and linked since they are waiting for their template to load asynchronously and their
+ * own compilation and linking has been suspended until that occurs. This hook can be considered
+ * analogous to the ngAfterViewInit and ngAfterContentInit hooks in Angular 2. Since the compilation
+ * process is rather different in Angular 1 there is no direct mapping and care should be taken when
+ * upgrading.
+ */
```
Comment (RedStarThrower) [inline src/component.ts:113]: Excellent comment

File: src/ng-register.ts [hunk 7]
```diff
@@ -44,7 +52,7 @@ class NgRegister {
       'filter',
       'controller',
       // 'directive',
-      'component',
+      // 'component',
```
Comment (RedStarThrower) [inline src/ng-register.ts:55]: Do we want to keep these comments here, for educational/illustration purposes?

File: example/src/index.ts [hunk 8]
```diff
@@ -1,15 +1,17 @@
...
 
-
 class DemoController {}
 
 ngRegister('DemoApp', [])
   .run(() => {
     console.log('Calling angular.module run function...');
   })
+  .component(ExComponent)
+  .component(ExRedComponent)
```
Comment (RedStarThrower) [inline example/src/index.ts:14]: ❤️

File: src/component.ts (hunk 4 above)
Comment (RedStarThrower) [reply src/component.ts]: Also, why the two underscores?

File: src/component.ts (hunk 6 above)
Comment (jmlopez-rod) [reply src/component.ts:113]: Ha, i wish those were my words, I just copied the angular docs for quick reference when we use these interfaces which they did not provide.

File: src/component.ts (hunk 5 above)
Comment (jmlopez-rod) [reply src/component.ts:65]: Yes, you get multiple controllers when you use `require`.

File: src/component.ts (hunk 4 above)
Comment (jmlopez-rod) [reply src/component.ts]: I'll fix this comment. The two underscores? Well, I wanted something that angular won't use, they already use `$inject`, I thought maybe using `$component` but that's kinda reserved for them so I just went for the `__`. I'm open to other prefix. Remember that this property is attached to the class constructor, i.e. is a static property.

File: src/ng-register.ts (hunk 7 above)
Comment (jmlopez-rod) [reply src/ng-register.ts:55]: That was the plan. I'm commenting out the methods I override.

File: example/src/ex-component.ts (hunk 1 above)
Comment (jmlopez-rod) [reply example/src/ex-component.ts:62]: No, we'll keep it simple and not use spaces surrounding braces. This is the code convention in airbnb.

**Review by RedStarThrower — APPROVED** [review]

**Pull request merged**